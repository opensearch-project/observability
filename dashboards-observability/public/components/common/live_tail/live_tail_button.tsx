/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

//Define pop over interval options for live tail button in your plugin

import { EuiButton } from "@elastic/eui";
import React, { useMemo } from "react";
import { LiveTailProps } from "common/types/explorer";

//Live Tail Button
export const LiveTailButton = ({
  isLiveTailOn,
  isLiveTailPopoverOpen,
  setIsLiveTailPopoverOpen,
  liveTailName,
  dataTestSubj,
}: LiveTailProps) => {
  const liveButton = useMemo(() => {
    return (
      <EuiButton
        iconType={isLiveTailOn ? "stop" : "play"}
        iconSide="left"
        onClick={() => setIsLiveTailPopoverOpen(!isLiveTailPopoverOpen)}
        data-test-subj={dataTestSubj}
      >
        {liveTailName}
      </EuiButton>
    );
  }, [isLiveTailPopoverOpen, isLiveTailOn]);
  return liveButton;
};

export const StopLiveButton = (props: any) => {
  const { StopLive, dataTestSubj } = props;

  const stopButton = () => {
    return (
      <EuiButton
        iconType="stop"
        onClick={() => StopLive()}
        color="danger"
        data-test-subj={dataTestSubj}
      >
        Stop
      </EuiButton>
    );
  };
  return stopButton();
};

export const sleep = (milliseconds: number | undefined) => {
  return new Promise((resolve) => setTimeout(resolve, milliseconds));
};
