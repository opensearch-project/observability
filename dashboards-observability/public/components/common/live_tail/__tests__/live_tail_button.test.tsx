/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { configure, mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { LiveTailButton, StopLiveButton } from '../live_tail_button';
import { waitFor } from '@testing-library/dom';

  describe('Live tail button', () => {
    configure({ adapter: new Adapter() });
  
    it('starts live tail with 5s interval', async () => {
      const setIsLiveTailPopoverOpen = jest.fn();

      const wrapper = mount(
        <LiveTailButton 
        isLiveTailOn={true} 
        setIsLiveTailPopoverOpen={setIsLiveTailPopoverOpen} 
        liveTailName={'5s'} 
        isLiveTailPopoverOpen={false} 
        dataTestSubj={''}   
        />
      );
      
      wrapper.update();
  
      await waitFor(() => {
        expect(wrapper).toMatchSnapshot();
      });
    });

    it('change live tail to 10s interval', async () => {
        const setIsLiveTailPopoverOpen = jest.fn();
        
        const wrapper = mount(
          <LiveTailButton 
          isLiveTailOn={true} 
          setIsLiveTailPopoverOpen={setIsLiveTailPopoverOpen} 
          liveTailName={'10s'} 
          isLiveTailPopoverOpen={false} 
          dataTestSubj={''}   
          />
        );
        
        wrapper.update();
    
        await waitFor(() => {
          expect(wrapper).toMatchSnapshot();
        });
      });
  });

  describe('Live tail off button', () => {
    configure({ adapter: new Adapter() });

    it('stop live tail', async () => {
        const StopLive = jest.fn();
        
        const wrapper = mount(
          <StopLiveButton 
             StopLive={StopLive}
             dataTestSubj={''}
          />
        );
        
        wrapper.update();
    
        await waitFor(() => {
          expect(wrapper).toMatchSnapshot();
        });
      });
 });