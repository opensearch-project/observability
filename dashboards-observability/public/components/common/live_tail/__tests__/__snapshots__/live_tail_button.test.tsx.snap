// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Live tail button change live tail to 10s interval 1`] = `
<LiveTailButton
  dataTestSubj=""
  isLiveTailOn={true}
  isLiveTailPopoverOpen={false}
  liveTailName="10s"
  setIsLiveTailPopoverOpen={[MockFunction]}
>
  <EuiButton
    data-test-subj=""
    iconSide="left"
    iconType="stop"
    onClick={[Function]}
  >
    <EuiButtonDisplay
      data-test-subj=""
      disabled={false}
      element="button"
      iconSide="left"
      iconType="stop"
      isDisabled={false}
      onClick={[Function]}
      type="button"
    >
      <button
        className="euiButton euiButton--primary"
        data-test-subj=""
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <EuiButtonContent
          className="euiButton__content"
          iconSide="left"
          iconType="stop"
          textProps={
            Object {
              "className": "euiButton__text",
            }
          }
        >
          <span
            className="euiButtonContent euiButton__content"
          >
            <EuiIcon
              className="euiButtonContent__icon"
              size="m"
              type="stop"
            >
              <EuiIconStop
                aria-hidden={true}
                className="euiIcon euiIcon--medium euiButtonContent__icon"
                focusable="false"
                role="img"
                style={null}
              >
                <svg
                  aria-hidden={true}
                  className="euiIcon euiIcon--medium euiButtonContent__icon"
                  focusable="false"
                  height={16}
                  role="img"
                  style={null}
                  viewBox="0 0 16 16"
                  width={16}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M4 2h8a2 2 0 012 2v8a2 2 0 01-2 2H4a2 2 0 01-2-2V4a2 2 0 012-2zm0 1a1 1 0 00-1 1v8a1 1 0 001 1h8a1 1 0 001-1V4a1 1 0 00-1-1H4z"
                    fillRule="evenodd"
                  />
                </svg>
              </EuiIconStop>
            </EuiIcon>
            <span
              className="euiButton__text"
            >
              10s
            </span>
          </span>
        </EuiButtonContent>
      </button>
    </EuiButtonDisplay>
  </EuiButton>
</LiveTailButton>
`;

exports[`Live tail button starts live tail with 5s interval 1`] = `
<LiveTailButton
  dataTestSubj=""
  isLiveTailOn={true}
  isLiveTailPopoverOpen={false}
  liveTailName="5s"
  setIsLiveTailPopoverOpen={[MockFunction]}
>
  <EuiButton
    data-test-subj=""
    iconSide="left"
    iconType="stop"
    onClick={[Function]}
  >
    <EuiButtonDisplay
      data-test-subj=""
      disabled={false}
      element="button"
      iconSide="left"
      iconType="stop"
      isDisabled={false}
      onClick={[Function]}
      type="button"
    >
      <button
        className="euiButton euiButton--primary"
        data-test-subj=""
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <EuiButtonContent
          className="euiButton__content"
          iconSide="left"
          iconType="stop"
          textProps={
            Object {
              "className": "euiButton__text",
            }
          }
        >
          <span
            className="euiButtonContent euiButton__content"
          >
            <EuiIcon
              className="euiButtonContent__icon"
              size="m"
              type="stop"
            >
              <EuiIconEmpty
                aria-hidden={true}
                className="euiIcon euiIcon--medium euiIcon-isLoading euiButtonContent__icon"
                focusable="false"
                role="img"
                style={null}
              >
                <svg
                  aria-hidden={true}
                  className="euiIcon euiIcon--medium euiIcon-isLoading euiButtonContent__icon"
                  focusable="false"
                  height={16}
                  role="img"
                  style={null}
                  viewBox="0 0 16 16"
                  width={16}
                  xmlns="http://www.w3.org/2000/svg"
                />
              </EuiIconEmpty>
            </EuiIcon>
            <span
              className="euiButton__text"
            >
              5s
            </span>
          </span>
        </EuiButtonContent>
      </button>
    </EuiButtonDisplay>
  </EuiButton>
</LiveTailButton>
`;

exports[`Live tail off button stop live tail 1`] = `
<StopLiveButton
  StopLive={[MockFunction]}
  dataTestSubj=""
>
  <EuiButton
    color="danger"
    data-test-subj=""
    iconType="stop"
    onClick={[Function]}
  >
    <EuiButtonDisplay
      color="danger"
      data-test-subj=""
      disabled={false}
      element="button"
      iconType="stop"
      isDisabled={false}
      onClick={[Function]}
      type="button"
    >
      <button
        className="euiButton euiButton--danger"
        data-test-subj=""
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <EuiButtonContent
          className="euiButton__content"
          iconSide="left"
          iconType="stop"
          textProps={
            Object {
              "className": "euiButton__text",
            }
          }
        >
          <span
            className="euiButtonContent euiButton__content"
          >
            <EuiIcon
              className="euiButtonContent__icon"
              size="m"
              type="stop"
            >
              <EuiIconStop
                aria-hidden={true}
                className="euiIcon euiIcon--medium euiButtonContent__icon"
                focusable="false"
                role="img"
                style={null}
              >
                <svg
                  aria-hidden={true}
                  className="euiIcon euiIcon--medium euiButtonContent__icon"
                  focusable="false"
                  height={16}
                  role="img"
                  style={null}
                  viewBox="0 0 16 16"
                  width={16}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M4 2h8a2 2 0 012 2v8a2 2 0 01-2 2H4a2 2 0 01-2-2V4a2 2 0 012-2zm0 1a1 1 0 00-1 1v8a1 1 0 001 1h8a1 1 0 001-1V4a1 1 0 00-1-1H4z"
                    fillRule="evenodd"
                  />
                </svg>
              </EuiIconStop>
            </EuiIcon>
            <span
              className="euiButton__text"
            >
              Stop
            </span>
          </span>
        </EuiButtonContent>
      </button>
    </EuiButtonDisplay>
  </EuiButton>
</StopLiveButton>
`;
